File: ai-newsroom/backend/app/task_status.py
from enum import Enum


class TaskStatus(str, Enum):
    PENDING = "pending"
    WRITING = "writing"
    WRITTEN = "written"
    REVIEWING = "reviewing"
    COMPLETED = "completed"
    FAILED = "failed"


TASK_STATUS_VALUES = tuple(status.value for status in TaskStatus)

LEGACY_TASK_STATUS_ALIASES = {
    "draft_ready": TaskStatus.WRITTEN.value,
    "in_progress": TaskStatus.WRITING.value,
}


def normalize_task_status(value: str | TaskStatus | None) -> str | None:
    if value is None:
        return None

    normalized = value.value if isinstance(value, TaskStatus) else str(value).strip().lower()
    if not normalized:
        return None

    if normalized in LEGACY_TASK_STATUS_ALIASES:
        return LEGACY_TASK_STATUS_ALIASES[normalized]

    if normalized in TASK_STATUS_VALUES:
        return normalized

    return None

File: ai-newsroom/frontend/src/lib/task-status.ts
export const TASK_STATUS_VALUES = [
  "pending",
  "writing",
  "written",
  "reviewing",
  "completed",
  "failed",
] as const;

export type TaskStatus = (typeof TASK_STATUS_VALUES)[number];
export type TaskBoardStage = "pending" | "writing" | "reviewing" | "completed" | "failed";
export type TaskLibraryBucket = "drafts" | "finalized" | null;

const LEGACY_TASK_STATUS_ALIASES: Record<string, TaskStatus> = {
  draft_ready: "written",
  in_progress: "writing",
};

const TASK_STATUS_SET = new Set<string>(TASK_STATUS_VALUES);

export function normalizeTaskStatus(status: string | null | undefined): TaskStatus {
  if (!status) {
    return "pending";
  }

  const normalized = status.trim().toLowerCase();
  if (normalized in LEGACY_TASK_STATUS_ALIASES) {
    return LEGACY_TASK_STATUS_ALIASES[normalized];
  }

  if (TASK_STATUS_SET.has(normalized)) {
    return normalized as TaskStatus;
  }

  return "pending";
}

export function getTaskBoardStage(status: string | null | undefined): TaskBoardStage {
  switch (normalizeTaskStatus(status)) {
    case "writing":
      return "writing";
    case "written":
    case "reviewing":
      return "reviewing";
    case "completed":
      return "completed";
    case "failed":
      return "failed";
    case "pending":
    default:
      return "pending";
  }
}

export function getTaskLibraryBucket(status: string | null | undefined): TaskLibraryBucket {
  switch (normalizeTaskStatus(status)) {
    case "writing":
    case "written":
    case "reviewing":
      return "drafts";
    case "completed":
      return "finalized";
    default:
      return null;
  }
}

export function isTaskVisibleInLibrary(status: string | null | undefined) {
  return getTaskLibraryBucket(status) !== null;
}

export function isEditorRecoverableTaskStatus(status: string | null | undefined) {
  const normalized = normalizeTaskStatus(status);
  return (
    normalized === "writing" ||
    normalized === "written" ||
    normalized === "reviewing" ||
    normalized === "completed" ||
    normalized === "failed"
  );
}
